File: leetcode/python/0482_license_key_formatting.py
class Solution:
    def licenseKeyFormatting(self, s: str, k: int) -> str:
        res = ""
        # j = 0
        # start from the back, using i loop through s:
        # if s[i] != "-":
        #   j += 1
        #   res += s[i]
        #   if j % k == 0 and i != 0: res += "-"
        if len(s) == 1 and s[0] != "-": return s.upper()
        elif len(s) == 1 and s[0] == "-": return ""

        j = 0
        for i in range(len(s) - 1, -1, -1):
            if s[i] != "-":
                j += 1
                res += s[i]
                if j % k == 0 and i != 0: res += "-"

        if res != "" and res[-1] == "-": res = res[:-1]
        
        return res.upper()[::-1]

File: leetcode/java/0482_license_key_formatting.java
class Solution {
    public String licenseKeyFormatting(String s, int k) {
        s = s.toUpperCase();
        StringBuilder res = new StringBuilder();
        // loop through s using i from s.length() - 1 to 0, j (0 to k):
        // res.append(s.charAt(i));
        // j++;
        // if (j == k && i != 0):
        //  res.append("-");
        // return res
        int lastChar = 0;
        for (int i = s.length() - 1, j = 0; i >= 0; i--) {
            if (s.charAt(i) != '-') {
                res.append(s.charAt(i));
                j++;
            }
            
            if (j == k && i != 0) {
                res.append("-");
                j = 0;
                lastChar = i;
            }
        }

        if (res.length() != 0 && res.charAt(res.length() - 1) == '-') res.deleteCharAt(res.length() - 1);

        return res.reverse().toString();
    }
}